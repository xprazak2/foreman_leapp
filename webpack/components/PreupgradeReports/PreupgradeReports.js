import React, { useState } from 'react';
import MessageBox from 'foremanReact/components/common/MessageBox';
import PropTypes from 'prop-types';
import { Button, Row, Spinner } from 'patternfly-react';

import { translate as __ } from 'foremanReact/common/I18n';

import './PreupgradeReports.scss';

import PreupgradeReportsList from '../PreupgradeReportsList';
import FixAlert from './components/FixAlert';

import {
  invocationFactory,
  flattenEntries,
  isEmpty,
  anyEntriesFixable,
  idsForInvocationFromReports,
  idsForInvocationFromEntries,
} from './PreupgradeReportsHelpers';

import { PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_SUCCESS } from '../../consts';

const PreupgradeReports = ({
  preupgradeReports,
  loading,
  error,
  onFixEntries,
  fixAllWorking,
  fixAllError,
}) => {
  if (!isEmpty(error)) {
    return (
      <MessageBox
        key="preupgrade-reports-error"
        icontype="error-circle-o"
        msg={`Could not retrieve data: ${error.statusText} - ${error.errorMsg}`}
      />
    );
  }

  const [checked, setChecked] = useState([]);

  const isSelected = entry => checked.some(item => item.id === entry.id);

  const anySelected = checked.length > 0;

  const toggleSelected = (entry, isEntrySelected) => {
    if (isEntrySelected) {
      setChecked(checked.filter(item => item.id !== entry.id));
    } else {
      setChecked([entry, ...checked]);
    }
  };

  const onFixAll = reports => {
    const ids = idsForInvocationFromReports(reports);

    onFix(ids);
  };

  const onFixSelected = () => {
    const ids = idsForInvocationFromEntries(checked);

    onFix(ids);
  };

  const onFix = async ids => {
    const action = await onFixEntries(invocationFactory(ids));

    if (action.type === PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_SUCCESS) {
      window.location.pathname = `/job_invocations/${action.payload.id}`;
    }
  };

  return (
    <React.Fragment>
      <Row>
        <div className="title-filter col-md-4">&nbsp;</div>
        <div id="title_action" className="col-md-8">
          <div className="btn-toolbar pull-right">
            {fixAllWorking && (
              <div id="toolbar-spinner">
                <Spinner loading size="sm" />
              </div>
            )}
            <Button
              onClick={() => onFixSelected()}
              disabled={!anyEntriesFixable(preupgradeReports) || fixAllWorking || !anySelected}
            >
              {__('Fix Selected')}
            </Button>
            <Button
              onClick={() => onFixAll(preupgradeReports)}
              disabled={!anyEntriesFixable(preupgradeReports) || fixAllWorking}
            >
              {__('Fix All')}
            </Button>
          </div>
        </div>
      </Row>
      {!isEmpty(fixAllError) && <FixAlert message={fixAllError.message} />}
      <PreupgradeReportsList
        allEntries={flattenEntries(preupgradeReports)}
        fixAllWorking={fixAllWorking}
        isSelected={isSelected}
        toggleSelected={toggleSelected}
      />
    </React.Fragment>
  );
};

PreupgradeReports.propTypes = {
  preupgradeReports: PropTypes.array.isRequired,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.object.isRequired,
  onFixEntries: PropTypes.func.isRequired,
  fixAllWorking: PropTypes.bool.isRequired,
  fixAllError: PropTypes.object.isRequired,
};

export default PreupgradeReports;
