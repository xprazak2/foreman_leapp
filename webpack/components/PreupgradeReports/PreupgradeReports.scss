@import '~@theforeman/vendor/scss/variables';

#leapp_preupgrade_report {
  .list-view-pf.working {
    .list-group-item.list-view-pf-stacked {
      color: $color-pf-black-400;

      &:hover{
        background-color: $color-pf-white;
        border-left-color: $color-pf-white;
        border-right-color: $color-pf-white;
      }
    }
  }
}
