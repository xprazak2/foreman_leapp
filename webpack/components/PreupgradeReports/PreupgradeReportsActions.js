import api from 'foremanReact/API';
import { deepPropsToCamelCase } from 'foremanReact/common/helpers';

import {
  PREUPGRADE_REPORTS_REQUEST,
  PREUPGRADE_REPORTS_SUCCESS,
  PREUPGRADE_REPORTS_FAILURE,
} from '../../consts';

export const getPreupgradeReports = url => async dispatch => {
  dispatch({ type: PREUPGRADE_REPORTS_REQUEST });

  try {
    const { data } = await api.get(url);
    return dispatch({
      type: PREUPGRADE_REPORTS_SUCCESS,
      payload: deepPropsToCamelCase(data),
    });
  } catch (error) {
    return dispatch(errorHandler(PREUPGRADE_REPORTS_FAILURE, error));
  }
};

const errorHandler = (msg, err) => {
  const error = {
    errorMsg: 'Failed to fetch preupgrade reports from server.',
    statusText: err.response.statusText,
  };
  return { type: msg, payload: { error } };
};
