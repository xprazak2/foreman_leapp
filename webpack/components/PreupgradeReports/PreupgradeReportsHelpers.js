export const flattenEntries = reports =>
  reports.reduce((memo, report) => [...memo, ...report.entries], []);

export const entryFixable = entry =>
  entry.detail &&
  entry.detail.remediations &&
  entry.detail.remediations.some(remediation => remediation.type === 'command');

export const isEmpty = obj => Object.keys(obj).length === 0;

export const anyEntriesFixable = reports =>
  flattenEntries(reports).some(entryFixable);

export const idsForInvocation = reports =>
  reports.reduce(
    (memo, report) => {
      report.entries.forEach(entry => {
        if (entryFixable(entry)) {
          memo.entryIds = [...memo.entryIds, entry.id];

          if (!memo.hostIds.includes(report.hostId)) {
            memo.hostIds = [...memo.hostIds, report.hostId];
          }
        }
      });
      return memo;
    },
    { hostIds: [], entryIds: [] }
  );
