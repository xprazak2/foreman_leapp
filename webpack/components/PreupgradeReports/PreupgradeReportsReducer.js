import Immutable from 'seamless-immutable';

import {
  PREUPGRADE_REPORTS_REQUEST,
  PREUPGRADE_REPORTS_SUCCESS,
  PREUPGRADE_REPORTS_FAILURE,
  PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_REQUEST,
  PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_SUCCESS,
  PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_FAILURE,
} from '../../consts';

export const initialState = Immutable({
  loadingPreupgradeReports: false,
  preupgradeReports: [],
  error: {},
  fixAllWorking: false,
  fixAllError: {},
});

const reducer = (state = initialState, action) => {
  const { payload } = action;
  switch (action.type) {
    case PREUPGRADE_REPORTS_REQUEST:
      return state.set('loadingPreupgradeReports', true);
    case PREUPGRADE_REPORTS_SUCCESS:
      return state.merge({
        loadingPreupgradeReports: false,
        preupgradeReports: payload,
      });
    case PREUPGRADE_REPORTS_FAILURE:
      return state.merge({
        error: payload.error,
        loadingPreupgradeReports: false,
      });
    case PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_REQUEST:
      return state.set('fixAllWorking', true);
    case PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_SUCCESS:
      return state.set('fixAllWorking', false);
    case PREUPGRADE_REPORTS_REMEDIATE_ENTRIES_FAILURE:
      return state.merge({
        fixAllWorking: false,
        fixAllError: { ...payload }, // what is payload???!
      });
    default:
      return state;
  }
};

export default reducer;
