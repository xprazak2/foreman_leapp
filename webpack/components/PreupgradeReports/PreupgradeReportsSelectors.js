export const selectPreupgrade = state => state.foremanLeapp.preupgrade;

export const selectPreupgradeReports = state => selectPreupgrade(state).preupgradeReports;
export const selectLoadingPreupgradeReports = state =>
  selectPreupgrade(state).loadingPreupgradeReports;
export const selectError = state => selectPreupgrade(state).error;
