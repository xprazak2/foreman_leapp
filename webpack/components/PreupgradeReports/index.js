import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';

import PreupgradeReports from './PreupgradeReports';

import {
  selectPreupgradeReports,
  selectLoadingPreupgradeReports,
  selectError,
} from './PreupgradeReportsSelectors';

import { getPreupgradeReports, postFixAll } from './PreupgradeReportsActions';

const WrappedPreupgradeReports = ({ url, jobInvocationsUrl }) => {
  const loading = useSelector(state => selectLoadingPreupgradeReports(state));
  const preupgradeReports = useSelector(state =>
    selectPreupgradeReports(state)
  );
  const error = useSelector(state => selectError(state));

  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getPreupgradeReports(url));
  }, [url]);

  const onFixEntries = (postData) => dispatch(postFixEntries(jobInvocationsUrl, postData));

  return (
    <PreupgradeReports
      preupgradeReports={preupgradeReports}
      error={error}
      loading={loading}
      onFixEntries={onFixEntries}
    />
  );
};

WrappedPreupgradeReports.propTypes = {
  url: PropTypes.string.isRequired,
};

export default WrappedPreupgradeReports;
