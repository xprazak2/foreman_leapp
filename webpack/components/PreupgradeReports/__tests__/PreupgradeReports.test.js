import { testComponentSnapshotsWithFixtures } from '@theforeman/test';

import PreupgradeReports from '../PreupgradeReports';

import { preupgradeReports } from './PreupgradeReports.fixtures';

const csrfToken = 'xyz';
const newJobInvocationUrl = '/job_invocations/new';

const fixtures = {
  'should render when loaded': {
    loading: false,
    error: {},
    preupgradeReports,
    csrfToken,
    newJobInvocationUrl,
  },
  'should render when loading': {
    loading: true,
    error: {},
    preupgradeReports: [],
    csrfToken,
    newJobInvocationUrl,
  },
  'should render error': {
    loading: false,
    error: {
      statusText: 'Internal server error',
      errorMsg: 'Well, this is embarassing',
    },
    preupgradeReports: [],
    csrfToken,
    newJobInvocationUrl,
  },
};

describe('PreupgradeReports', () =>
  testComponentSnapshotsWithFixtures(PreupgradeReports, fixtures));
