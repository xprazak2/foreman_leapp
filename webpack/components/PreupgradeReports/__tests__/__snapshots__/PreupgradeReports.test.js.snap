// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreupgradeReports should render error 1`] = `
<MessageBox
  icontype="error-circle-o"
  key="preupgrade-reports-error"
  msg="Could not retrieve data: Internal server error - Well, this is embarassing"
/>
`;

exports[`PreupgradeReports should render when loaded 1`] = `
<LoadingState
  additionalClasses=""
  loading={false}
  loadingText="Loading"
  size="lg"
  timeout={300}
>
  <Row
    bsClass="row"
    componentClass="div"
  >
    <div
      className="col-md-8"
    >
      <EntriesFilter
        filterType="title"
        filterValue=""
        onFilterTypeChange={[Function]}
        onFilterValueChange={[Function]}
      />
    </div>
    <div
      className="col-md-4"
    >
      <div
        className="btn-toolbar pull-right"
      >
        <FixSelectedButton
          csrfToken="xyz"
          disabled={true}
          ids={
            Object {
              "entryIds": Array [],
              "hostIds": Array [],
            }
          }
          postUrl="/job_invocations/new"
        />
      </div>
    </div>
  </Row>
  <PreupgradeReportsList
    allEntries={
      Array [
        Object {
          "hostname": "host.example.com",
          "id": 42,
          "severity": "Too severe to talk about",
          "title": "Fix me!",
        },
        Object {
          "hostname": "host.example.com",
          "id": 43,
          "severity": "medium",
          "title": "I am broken too",
        },
        Object {
          "hostname": "foo.example.com",
          "id": 44,
          "severity": "high",
          "title": "Octocat is not happy",
        },
        Object {
          "hostname": "foo.example.com",
          "id": 45,
          "severity": "low",
          "title": "Not enough credits",
        },
        Object {
          "hostname": "foo.example.com",
          "id": 46,
          "severity": "medium",
          "title": "SELinux is turned off",
        },
        Object {
          "hostname": "foo.example.com",
          "id": 47,
          "severity": "medium",
          "title": "Root password is too short",
        },
        Object {
          "hostname": "foo.example.com",
          "id": 49,
          "severity": "high",
          "title": "No chocolate chip cookies in cookie jar",
        },
      ]
    }
    isSelected={[Function]}
    toggleSelected={[Function]}
  />
</LoadingState>
`;

exports[`PreupgradeReports should render when loading 1`] = `
<LoadingState
  additionalClasses=""
  loading={true}
  loadingText="Loading"
  size="lg"
  timeout={300}
>
  <Row
    bsClass="row"
    componentClass="div"
  >
    <div
      className="col-md-8"
    >
      <EntriesFilter
        filterType="title"
        filterValue=""
        onFilterTypeChange={[Function]}
        onFilterValueChange={[Function]}
      />
    </div>
    <div
      className="col-md-4"
    >
      <div
        className="btn-toolbar pull-right"
      >
        <FixSelectedButton
          csrfToken="xyz"
          disabled={true}
          ids={
            Object {
              "entryIds": Array [],
              "hostIds": Array [],
            }
          }
          postUrl="/job_invocations/new"
        />
      </div>
    </div>
  </Row>
  <PreupgradeReportsList
    allEntries={Array []}
    isSelected={[Function]}
    toggleSelected={[Function]}
  />
</LoadingState>
`;
